### PR TITLE
Fix errors in eip712 signing

### LINKED ISSUE #149: encode_structured_data Error with Array Types (SignTypedDataV4)
## What was wrong?

`encode_structured_data` throws KeyError when there is an array type in a struct. The error doesn't give any context and seems to support some array types (ex. `Person[]` doesnt seem to throw error but `address[]` throws error).  When I simply just make the Person.wallets to be `address` instead of `address[]` it works (obviously I update the message accordingly). But for some reason, an array of address just is not working. For more context, I am trying to make it work with the SignTypedDataV4 on MetaMasks[ test Dapp](https://metamask.github.io/test-dapp/)

### Code that produced the error

```py
msg = {
      'domain': {
        'chainId': 1,
        'name': 'Ether Mail',
        'verifyingContract': '0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC',
        'version': '1',
      },
      'message': {
        'contents': 'Hello, Bob!',
        'from': {
          'name': 'Cow',
          'wallets': ['0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826',
                      '0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826'],
        },
        'to': [
          {
            'name': 'Bob',
            'wallets': ['0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826',
                        '0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826'],
          },
        ],
      },
      'primaryType': 'Mail',
      'types': {
        'EIP712Domain': [
          { 'name': 'name', 'type': 'string' },
          { 'name': 'version', 'type': 'string' },
          { 'name': 'chainId', 'type': 'uint256' },
          { 'name': 'verifyingContract', 'type': 'address' },
        ],
        'Group': [
          { 'name': 'name', 'type': 'string' },
          { 'name': 'members', 'type': 'Person[]' },
        ],
        'Mail': [
          { 'name': 'from', 'type': 'Person' },
          { 'name': 'to', 'type': 'Person[]' },
          { 'name': 'contents', 'type': 'string' },
        ],
        'Person': [
          { 'name': 'name', 'type': 'string' },
          { 'name': 'wallets', 'type': 'address[]' },
        ],
      },
    }
  
    new_msg = json.loads(json.dumps(msg), object_hook=_decode)
    new_msg['domain']['version'] = str(new_msg['domain']['version'])
    encoded_msg = encode_structured_data(new_msg)
    signed_msg = w3.eth.account.sign_message(encoded_msg, account)
```

### Full error output 

```sh
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/user/sign_message.py", line 158, in t 
    encoded_msg = encode_structured_data(new_msg)
  File "/home/user/.local/lib/python3.8/site-packages/eth_account/messages.py", line 156, in encode_structured_data
    hash_eip712_message(structured_data),
  File "/home/user/.local/lib/python3.8/site-packages/eth_account/_utils/structured_data/hashing.py", line 286, in hash_message
    encode_data(
  File "/home/user/.local/lib/python3.8/site-packages/eth_account/_utils/structured_data/hashing.py", line 262, in encode_data
    data_types_and_hashes = _encode_data(primaryType, types, data)
  File "/home/user/.local/lib/python3.8/site-packages/eth_utils/functional.py", line 45, in inner
    return callback(fn(*args, **kwargs))
  File "/home/user/.local/lib/python3.8/site-packages/eth_account/_utils/structured_data/hashing.py", line 209, in _encode_data
    hashed_value = keccak(primitive=encode_data(field["type"], types, value))
  File "/home/user/.local/lib/python3.8/site-packages/eth_account/_utils/structured_data/hashing.py", line 262, in encode_data
    data_types_and_hashes = _encode_data(primaryType, types, data)
  File "/home/user/.local/lib/python3.8/site-packages/eth_utils/functional.py", line 45, in inner
    return callback(fn(*args, **kwargs))
  File "/home/user/.local/lib/python3.8/site-packages/eth_account/_utils/structured_data/hashing.py", line 232, in _encode_data
    array_items_encoding = [
  File "/home/user/.local/lib/python3.8/site-packages/eth_account/_utils/structured_data/hashing.py", line 233, in <listcomp>
    encode_data(parsed_type.base, types, array_item)
  File "/home/user/.local/lib/python3.8/site-packages/eth_account/_utils/structured_data/hashing.py", line 262, in encode_data
    data_types_and_hashes = _encode_data(primaryType, types, data)
  File "/home/user/.local/lib/python3.8/site-packages/eth_utils/functional.py", line 45, in inner
    return callback(fn(*args, **kwargs))
  File "/home/user/.local/lib/python3.8/site-packages/eth_account/_utils/structured_data/hashing.py", line 174, in _encode_data
    yield "bytes32", hash_struct_type(primary_type, types)
  File "/home/user/.local/lib/python3.8/site-packages/eth_account/_utils/structured_data/hashing.py", line 96, in hash_struct_type
    return keccak(text=encode_type(primary_type, types))
  File "/home/user/.local/lib/python3.8/site-packages/eth_account/_utils/structured_data/hashing.py", line 83, in encode_type
    deps = get_dependencies(primary_type, types)
  File "/home/user/.local/lib/python3.8/site-packages/eth_account/_utils/structured_data/hashing.py", line 40, in get_dependencies
    fields = types[struct_name]
KeyError: 'address'
```

### Environment

```sh
Python version:
3.8.10 (default, Nov 26 2021, 20:14:08) 
[GCC 9.3.0]

Operating System: Linux-5.10.16.3-microsoft-standard-WSL2-x86_64-with-glibc2.29

pip freeze result:
aiohttp==3.8.1
aiosignal==1.2.0
async-timeout==4.0.2
attrs==19.3.0
Automat==0.8.0
base58==2.1.1
bitarray==1.2.2
blinker==1.4
certifi==2019.11.28
chardet==3.0.4
charset-normalizer==2.0.12
click==8.0.4
cloud-init==20.1
colorama==0.4.3
command-not-found==0.3
configobj==5.0.6
constantly==15.1.0
cryptography==2.8
cytoolz==0.11.2
dbus-python==1.2.16
distro==1.4.0
distro-info===0.23ubuntu1
dnspython==2.2.0
entrypoints==0.3
eth-abi==3.0.0
eth-account==0.6.1
eth-hash==0.3.2
eth-keyfile==0.6.0
eth-keys==0.4.0
eth-rlp==0.3.0
eth-typing==3.0.0
eth-utils==2.0.0
Flask==2.0.1
Flask-Cors==3.0.10
frozenlist==1.3.0
hexbytes==0.2.2
httplib2==0.14.0
hyperlink==19.0.0
idna==2.8
importlib-metadata==1.5.0
incremental==16.10.1
ipfshttpclient==0.8.0a2
itsdangerous==2.1.0
Jinja2==3.0.3
jsonpatch==1.22
jsonpointer==2.0
jsonschema==3.2.0
keyring==18.0.1
language-selector==0.1
launchpadlib==1.10.13
lazr.restfulclient==0.14.2
lazr.uri==1.0.3
lru-dict==1.1.7
MarkupSafe==2.1.0
more-itertools==4.2.0
multiaddr==0.0.9
multidict==6.0.2
netaddr==0.8.0
netifaces==0.10.4
oauthlib==3.1.0
parsimonious==0.8.1
protobuf==3.19.4
pyasn1==0.4.2
pyasn1-modules==0.2.1
pycryptodome==3.14.1
PyGObject==3.36.0
PyHamcrest==1.9.0
PyJWT==1.7.1
pymacaroons==0.13.0
pymongo==3.11.4
PyNaCl==1.3.0
pyOpenSSL==19.0.0
pyrsistent==0.15.5
pyserial==3.4
python-apt==2.0.0
python-debian===0.1.36ubuntu1
python-dotenv==0.19.2
PyYAML==5.3.1
requests==2.25.1
requests-unixsocket==0.2.0
rlp==3.0.0
SecretStorage==2.3.1
service-identity==18.1.0
simplejson==3.16.0
six==1.16.0
ssh-import-id==5.10
stripe==2.66.0
systemd-python==234
toolz==0.11.2
Twisted==18.9.0
ubuntu-advantage-tools==20.3
ufw==0.36
unattended-upgrades==0.1
urllib3==1.25.8
varint==1.0.2
wadllib==1.3.3
web3==5.28.0
websockets==9.1
Werkzeug==2.0.3
yarl==1.7.2
zipp==1.0.0
zope.interface==4.7.1
``` 

## How can it be fixed?

Unsure of how to solve it

### LINKED ISSUE #152: sign-typed-data-v4  hope add support for arrays
hope add support for arrays thanks
 https://docs.metamask.io/guide/signing-data.html#sign-typed-data-v4
 https://medium.com/metamask/eip712-is-coming-what-to-expect-and-how-to-use-it-bb92fd1a7a26

### LINKED ISSUE #174: Fix errors in EIP-712 signing
## What was wrong?

I noticed that eth-account calculates EIP-712 signatures differently from Metamask under these scenarios:

1. When there is an array in the message, array contents are not keccak()ed.
2. When there is an array type, type dependencies are not added to encode_type()

### Code that produced the error

Here's a micro example:
```js
const { TypedDataUtils } = require("@metamask/eth-sig-util/sign-typed-data")

const data = require("./tests/fixtures/valid_eip712_example_with_array.json")
const encodedData = TypedDataUtils.encodeData(data.primaryType, data.message, data.types, "V4")
const hashedStruct = TypedDataUtils.hashStruct(data.primaryType, data.message, data.types, "V4")
const eip712Hash = TypedDataUtils.eip712Hash(data, "V4")

// Metamask:
// "b19c5c7781083ed7325660d847f8c547b744917d67b38858b00bce7e62a4866bfc71e5fa27ff56c350aa"
// "531bc129ebdf613b772b6604664f5d8dbe21b85eb0c8cd54f074a4af31b4411ff6a60c9719dbd559c221"
// "c8ac3492d9d872b041d703d105a7f44bc01de4515b63df31009d8ce4fe8b06eb06a1ff4eec0b562d1c70"
// "3535b5aadf3154a261abdd9086fc627b61efca26ae5702701d05cd2305f7c52a2fc8"
// eth-account:
// "b19c5c7781083ed7325660d847f8c547b744917d67b38858b00bce7e62a4866bfc71e5fa27ff56c350aa"
// "531bc129ebdf613b772b6604664f5d8dbe21b85eb0c8cd54f074a4af31b4411ff6a60c9719dbd559c221"
// "c8ac3492d9d872b041d703d1a88b5e35e8f6bccc3d72ef467ed581615888f8cf34df19d068d04225439b"
// "e65ab5aadf3154a261abdd9086fc627b61efca26ae5702701d05cd2305f7c52a2fc8"
console.log("encodeData valid_eip712_example_with_array:")
console.log(encodedData.toString("hex"))

// Metamask: 04e58c373e4cd0695f0c2d6e2d8e379d6b51b987cf27d582dd94a03532d430d3
// eth-account: 76649452daa3e4101beef5b01669fd0de45ece35188d529703c73526a2454521
console.log("hashStruct valid_eip712_example_with_array:")
console.log(hashedStruct.toString("hex"))

// Metamask: 1780e7e042fa9ec126ccb68cd707d61580d00601b3eff8a5ec05116b46007fdb
// eth-account: 4e3c4173651b3054c0635dfae57a3b65ae1527ed0ba9ff76cecb6d9807687a7f
console.log("eip712Hash valid_eip712_example_with_array:")
console.log(eip712Hash.toString("hex"))
```

## How can it be fixed?

Fixes to come in PR.

### LINKED ISSUE #180: EIP-712 Signatures generated by `eth-account` and `ether-js` are different when the message contains arrays
## What was wrong?

Following issue #174, I noticed that signatures produced by `eth-account` are not matching ether-js's [`_signTypedData`](https://docs.ethers.io/v5/api/signer/#Signer-signTypedData)  for the example message with arrays [valid_eip712_example_with_array.json](https://github.com/ethereum/eth-account/blob/c35375a0f2441b6350df8fe5b57c22f1bcc8c52b/tests/fixtures/valid_eip712_example_with_array.json).

### Code that produced the error

Generate signature with `eth-account`:

```py
import os
import json
from eth_account.messages import encode_structured_data
from eth_account import Account

pk = os.environ['PRIVATE_KEY']

structured_msg = encode_structured_data(json.loads(open("valid_eip712_example_with_array.json", "r").read()))
signed = Account.sign_message(structured_msg, pk)
new_addr = Account.recover_message(structured_msg, signed.signature)
print(signed.signature.hex())
```

Generate signature with `ether-js`:

```js
const ethers = require('ethers')

var domainData =
  {
    "name": "Ether Mail",
    "version": "1",
    "chainId": 1,
    "verifyingContract": "0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC"
  }

var message = {
  "from": {
    "name": "Cow",
    "wallet": "0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826"
  },
  "to": {
    "name": "Bob",
    "wallet": "0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB"
  },
  "cc": [
    {
      "name": "Alice",
      "wallet": "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
    },
    {
      "name": "Dot",
      "wallet": "0xdddddddddddddddddddddddddddddddddddddddd"
    }
  ],
  "contents": "Hello, Bob!"
}

var EIP_712_ORDER_TYPE = {
  "Person": [
    {"name": "name", "type": "string"},
    {"name": "wallet", "type": "address"}
  ],
  "Mail": [
    {"name": "from", "type": "Person"},
    {"name": "to", "type": "Person"},
    {"name": "cc", "type": "Person[]"},
    {"name": "contents", "type": "string"}
  ]
}

var signer = new ethers.Wallet(PRIVATE_KEY);

var signature = await signer._signTypedData(
  domainData,
  EIP_712_ORDER_TYPE,
  message
);

console.log(signature)
```

### Expected Result

I expected both signatures to match, like they did for the example [valid_eip712_example.json](https://github.com/ethereum/eth-account/blob/c35375a0f2441b6350df8fe5b57c22f1bcc8c52b/tests/fixtures/valid_eip712_example.json).

### Environment

```
eth-account==0.6.1
ethers/5.6.9
```

### PR DESCRIPTION
**Pull request by doubleukay**
## What was wrong?

closes #149 
closes #152 
closes #174
closes #180

## How was it fixed?

@doubleukay:
- Opened up the logic to accept arrays with more than one dimension.
- Fixed expected hexes in tests for cases we were already testing.

---
@pacrob:
- Bulk of the work in fixing the logic here, while comparing to `eth-sig-utils` (Metamask). This included fixing the logic so that "atomic" value arrays are also correct.
- Added bulk of the tests, compared against `eth-sig-utils` (Metamask).

---
@fselmo:

- Invert the dimensions that we extract from arrays via `get_array_dimensions()` since this is how the ABI, Solidity expects them. Compared with `eth-sig-utils` as well.
- Add `"dynamic"` dimension when varying length arrays are used (dynamic).
- Added tests for correct array dimension order and for "dynamic" dimensions.
- Added to ignored `pydocstyle` rules.

### To-Do

- [ ] **Clean up commit history**
- [x] Add entry to the [release notes](https://github.com/ethereum/eth-account/blob/master/newsfragments/README.md)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Review by Gruzin733 — APPROVED** [review]

**Review by and00999 — APPROVED** [review]

**Comment by pacrob** [comment]
Sorry for the delay on this, but I'm still checking edge cases. This fix does provide the same output as Metamask for a single-depth array of a user-defined type, but chokes on single-depth arrays of `string` or `uint256`, for example. And the outputs diverge on multi-dimensional arrays.

Thanks for your patience! I'll have it out asap.

**Comment by pacrob** [comment]
The main functionality and error messaging has been added. I'll do more testing tomorrow.

File: eth_account/_utils/structured_data/hashing.py
Comment (fselmo) [inline eth_account/_utils/structured_data/hashing.py:187]: I noticed a few issues here when dynamic arrays are used along with defined length arrays. `x[0]` will throw an "index out of bounds" error since it's an empty tuple for dynamic array dimensions. I'm going to think a bit more on this one as I think it has led to some other test cases I want to test out as well. It would be useful to compare the values accepted here to values accepted in Solidity and make sure they match as I think I'm seeing some discrepancies with multi arrays and validation errors.

**Review by fselmo — COMMENTED** [review]
Thanks for all the hard work here @doubleukay and @pacrob! This definitely needed this kind of attention. I think this is in pretty good shape now.

I did a first pass through and have mostly cosmetic comments for now though I left a comment about testing nested tuple types against those accepted by Solidity / ABI encoding. 

There are some cases I still want to test the validation for as I think I'm seeing a bit of discrepancy on whether they are accepted types. In particular, dynamic dimension with defined dimension, i.e. `uint256[][3]` or `uint256[3][]`, etc... I'll try to get to testing this asap but I think we're really close with this.

**Comment by fselmo** [comment]
@doubleukay @pacrob thanks again for taking the time to get this moving! This is in pretty high in demand especially as of late.

I added a commit to address the test cases I was seeing, from my review yesterday, when comparing with how solidity accepts nested array types. It seems we had it inverted in the library, which was fine in the past since we purposefully at some point decided to keep all arrays to a single dimension. Now that we opened it up to multiple dimensions, it appeared to be inverted.

It's important to note all tests added here by @pacrob and the tweaks I made to them were tested against the current `eth-sig-utils` which is what Metamask is using under the hood and they all match.

I don't see too much at the moment except I'd like to eventually look into some ABI types that might be possible to define in the schema that may not be acceptable array types... but this is a huge improvement on what was already here, especially towards EIP-712 / signTypedData_V4 compatibility.

Anything else you can think of that might be useful to include here? Any more tests you can think of?

**Review by fselmo — APPROVED** [review]
This looks good to me. It would be good to get one more set of eyes on it before merging so I'd like to wait for a @kclowes approval first though. Whenever you get a chance :eyes:

File: eth_account/_utils/structured_data/hashing.py
Comment (kclowes) [inline eth_account/_utils/structured_data/hashing.py:52]: Since one of the bugs was that we were skipping over an array where we didn't know what it was, I wonder if we should make this an `elif` and raise on the else? Or if we could even define that condition?

**Review by kclowes — APPROVED** [review]
This looks good to me! Thanks to everyone involved! I had a few nits and a few questions, but feel free to change or keep whatever you see fit. I don't think the questions are serious enough to block merging for a whole week. :shipit: !

File: eth_account/_utils/structured_data/hashing.py
Comment (pacrob) [reply eth_account/_utils/structured_data/hashing.py:146]: deleted

File: tests/core/test_structured_data_signing.py
Comment (pacrob) [reply tests/core/test_structured_data_signing.py:550]: fixed

File: tests/core/test_structured_data_signing.py
Comment (pacrob) [reply tests/core/test_structured_data_signing.py:758]: fixed

File: tests/core/test_structured_data_signing.py
Comment (pacrob) [reply tests/core/test_structured_data_signing.py:733]: fixed

File: eth_account/_utils/structured_data/hashing.py
Comment (fselmo) [reply eth_account/_utils/structured_data/hashing.py:142]: > I'm pretty sure we were grabbed smallest dimension of the array (the first dim in a sorted order) and returned that for the dimension at that depth.

Actually, this would've been true if we even accepted more than one dimension but we were validating that only one dimension was accepted. So I suppose there used to be validation there but now we properly accept varying dimensions and label that depth as "dynamic". Wasn't sure what else would be appropriate here. There may be some better messaging we can provide either now or in the future but at the moment this should open it up to accept them.

**Pull request merged**
